controllers/utils/base_reconciler.go

```diff
@@ -0,0 +1,408 @@
+	return ctrl.Result{}, nil
+}
+
+func (r *ReconciliationRunner) IfDeletionTimestampIsZero(fn CloudStackReconcilerMethod) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		if r.ConditionalResult = r.ReconciliationSubject.GetDeletionTimestamp().IsZero(); r.ConditionalResult {
+			return fn()
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+func (r *ReconciliationRunner) Else(fn CloudStackReconcilerMethod) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		if !r.ConditionalResult {
+			return fn()
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+// GetCAPICluster gets the CAPI cluster the reconciliation subject belongs to.
+func (r *ReconciliationRunner) GetCAPICluster() (ctrl.Result, error) {
+	name := r.ReconciliationSubject.GetLabels()[capiv1.ClusterLabelName]
+	if name == "" {
+		r.Log.V(1).Info("Reconciliation Subject is missing cluster label or cluster does not exist. Skipping CAPI Cluster fetch.",
+			"SubjectKind", r.ReconciliationSubject.GetObjectKind().GroupVersionKind().Kind)
+		return ctrl.Result{}, nil
+	}
+	r.CAPICluster = &capiv1.Cluster{}
+	key := client.ObjectKey{
+		Namespace: r.ReconciliationSubject.GetNamespace(),
+		Name:      name,
+	}
+	if err := r.Client.Get(r.RequestCtx, key, r.CAPICluster); err != nil {
+		return ctrl.Result{}, errors.Wrap(client.IgnoreNotFound(err), "error encountered while getting CAPI Cluster "+name)
+	} else if r.CAPICluster.Name == "" {
+		return r.RequeueWithMessage("Cluster not fetched.")
+	}
+	return ctrl.Result{}, nil
+}
+
+// GetCSCluster gets the CAPI cluster the reconciliation subject belongs to.
+func (r *ReconciliationRunner) GetCSCluster() (ctrl.Result, error) {
+	name := r.ReconciliationSubject.GetLabels()[capiv1.ClusterLabelName]
+	if name == "" {
+		r.Log.V(1).Info("Reconciliation Subject is missing cluster label or cluster does not exist. Skipping CloudStackCluster fetch.",
+			"SubjectKind", r.ReconciliationSubject.GetObjectKind().GroupVersionKind().Kind)
+		return ctrl.Result{}, nil
+	}
+	r.CSCluster = &infrav1.CloudStackCluster{}
+	key := client.ObjectKey{
+		Namespace: r.ReconciliationSubject.GetNamespace(),
+		Name:      name,
+	}
+	err := r.Client.Get(r.RequestCtx, key, r.CSCluster)
+	return ctrl.Result{}, errors.Wrap(client.IgnoreNotFound(err), "error encountered while getting CAPI Cluster "+name)
+}
+
+// CheckOwnedCRDsForReadiness queries for the readiness of CRDs of GroupVersionKind passed.
+func (r *ReconciliationRunner) CheckOwnedCRDsForReadiness(gvks ...schema.GroupVersionKind) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		// For each GroupVersionKind...
+		for _, gvk := range gvks {
+			// Query to find objects of this kind.
+			// TODO: Filter use owner label to filter. Will need to build generic owner labeling system too.
+			potentiallyOnwedObjs := &unstructured.UnstructuredList{}
+			potentiallyOnwedObjs.SetGroupVersionKind(gvk)
+			err := r.Client.List(r.RequestCtx, potentiallyOnwedObjs)
```

Comment:
Does this have potential to retrieve a large number of objects?  If so, any concerns?